src/tokens/ERC20Gauges.sol

```diff
@@ -407,43 +276,40 @@ abstract contract ERC20Gauges is ERC20 {
     */
     function decrementGauge(
         address gauge,
-        uint112 weight
-    ) public virtual returns (uint112 newUserWeight) {
-        uint32 currentCycle = _getGaugeCycleEnd();
-
+        uint256 weight
+    ) public virtual returns (uint256 newUserWeight) {
         // All operations will revert on underflow, protecting against bad inputs
-        _decrementGaugeWeight(msg.sender, gauge, weight, currentCycle);
-        return _decrementUserAndGlobalWeights(msg.sender, weight, currentCycle);
+        _decrementGaugeWeight(msg.sender, gauge, weight);
+        totalTypeWeight[gaugeType[gauge]] -= weight;
```

Comment:
I'm curious why there is this addition of types

Reply:
shouldn't this line be in the `_decrementGaugeWeight` function?

Reply:
Maybe we should reorganize the code but there is a reason why this isn't in the `_decrementGaugeWeight` function, this internal function is not called everytime the type weight must decrease (it is not called in `_removeGauge`, because this is a global weight and it is not linked to a user). I initially added this in the `_decrementGaugeWeight` function, but then it would revert on `_decrementWeightUntilFree` because the `totalTypeWeight` is already decremented when we deprecate a gauge (`_removeGauge`), and it would try to decrement the type weight a second time. 

I think the current implementation decrements all values properly at the right time, here is a recap : 

![image](https://github.com/volt-protocol/ethereum-credit-guild/assets/6916865/21cd145a-fb56-4a7c-b087-537556a60bba)

Reply:
Rational for adding multiple gauge types is to support multiple CREDIT tokens in the future, each with their instance of the protocol, each with their gauges and their per-type allocation of debt ceiling, but all these instances can share the same GUILD token.

Reply:
so the reason to keep things out of those individual functions is the gas savings you get from the single write to storage after decrementWeightUntilFree loop completes?

seems that they both call decrementGuage weight and getUserWeight could be decremented there for this variable.

Reply:
Yep, I agree it's less efficient to do multiple SSTORE on `getUserWeight` but it's worth the code simplification to move it into `_decremenGaugeWeight`.

Looking at the above colored example, I also saw a bug & added a test to catch it : in `_decrementWeightUntilFree`, `totalWeight` (and my added `totalTypeWeight`) is decremented only if the gauge is not deprecated (`_removeGauge` already decrements total & type weight, so this would decrement twice if we were decrementing `totalWeight` & `totalTypeWeight` for deprecated gauges). The regular `decrementGauge(s)` function does not do this check, and can revert due to underflow from decrementing `totalWeight` twice (when decrementing votes on a deprecated gauge).

Reply:
![image](https://github.com/volt-protocol/ethereum-credit-guild/assets/6916865/c0948878-f262-429b-acd0-4bcb2f04db5d)
ok it's much cleaner now 😄 
I couldn't factorize `totalWeight--` and `totalTypeWeight--` in a common internal function, because on the image they seem like they are called exactly at the same time (because I visually grouped them), but in the code the `totalWeight--` are actually batched for functions that have a loop (`decrementGauges`, `_decrementWeightUntilFree`) to do a singular SSTORE, whereas in functions without loop (`decrementGauge`, `_removeGauge`) total & type weight decrease are happening at the same time.